packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+        uint256 _minReturnAmountAfterFee,
+        uint256 _lockPeriodInSec
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 amount_, uint256 unlockTimeInSec_)
+    {
+        moxieToken.safeTransferFrom(msg.sender, address(this), _depositAmount);
+        moxieToken.approve(address(moxieBondingCurve), _depositAmount);
+        amount_ = moxieBondingCurve.buyShares(
+            _subject,
+            _depositAmount,
+            _minReturnAmountAfterFee
+        );
+
+        (, unlockTimeInSec_) = _createLock(
+            _subject,
+            amount_,
+            _lockPeriodInSec,
+            msg.sender,
+            true
+        );
+    }
+
+    /**
+     * External function to buy & lock tokens.
+     * @param _subject Subject address for which tokens are being bought & deposited.
+     * @param _depositAmount amount of moxie tokens getting deposited.
+     * @param _minReturnAmountAfterFee Slippage setting which determines minimum amount of tokens after fee.
+     * @param _lockPeriodInSec Lock period for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @return amount_ Amount of tokens bought & locked.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function buyAndLockFor(
+        address _subject,
+        uint256 _depositAmount,
+        uint256 _minReturnAmountAfterFee,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 amount_, uint256 unlockTimeInSec_)
+    {
+        moxieToken.safeTransferFrom(msg.sender, address(this), _depositAmount);
+        moxieToken.approve(address(moxieBondingCurve), _depositAmount);
+        amount_ = moxieBondingCurve.buyShares(
+            _subject,
+            _depositAmount,
+            _minReturnAmountAfterFee
+        );
+        (, unlockTimeInSec_) = _createLock(
+            _subject,
+            amount_,
+            _lockPeriodInSec,
+            _beneficiary,
+            true
+        );
+    }
+
+    /**
+     * External function to buy & lock multiple tokens.
+     * @param _subjects Subject addresses for which tokens are being bought & deposited.
+     * @param _depositAmounts Amounts of moxie tokens getting deposited.
+     * @param _minReturnAmountsAfterFee Slippage settings which determine minimum amounts of tokens after fee.
+     * @param _lockPeriodInSec Lock periods for the tokens.
+     * @return amounts_ Amounts of tokens bought & locked.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function buyAndLockMultiple(
+        address[] memory _subjects,
+        uint256[] memory _depositAmounts,
+        uint256[] memory _minReturnAmountsAfterFee,
+        uint256 _lockPeriodInSec
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256[] memory amounts_, uint256 unlockTimeInSec_)
```

Comment:
unlockTimeInSec_ this should array as well

Reply:
All tokens are locked for single duration. so only one unlock time.